spec/fastsync/version2/scheduler.tla

```diff
@@ -0,0 +1,580 @@
+(* ----------------------------------------------------------------------------------------------*)
+InitSc ==
+  /\ scRunning = TRUE
+  /\ outEvent = noEvent
+  /\ \E startHeight \in Heights:
+    scheduler = [
+      initHeight |-> startHeight,
+      height |-> startHeight,
+      peers |-> {},
+      peerHeights |-> [p \in PeerIDs |-> None],
+      peerStates |-> [p \in PeerIDs |-> "peerStateUnknown"],
+      blocks |-> {},
+      blockStates |-> [h \in Heights |-> "blockStateUnknown"],
+      pendingBlocks |-> [h \in Heights |-> None],
+      receivedBlocks |-> [h \in Heights |-> None]
+      ]
+     
+handleAddNewPeer ==
+  /\ inEvent.type = "bcAddNewPeer"
+  /\ LET res == addPeer(scheduler, inEvent.peerID) IN
+     IF res.err /= noErr THEN
+       /\ outEvent' = [type |-> "scSchedulerFail", reason |-> res.err]
+       /\ UNCHANGED <<scheduler>>
+     ELSE
+       /\ scheduler' = res.val
+       /\ UNCHANGED outEvent
+  /\ UNCHANGED scRunning
+       
+handleRemovePeer ==
+  /\ inEvent.type = "bcRemovePeer"
+  /\ LET res == removePeer(scheduler, inEvent.peerID) IN
+     IF res.err /= noErr THEN
+       /\ outEvent' = [type |-> "scSchedulerFail", reason |-> res.err]
+       /\ UNCHANGED scheduler
+     ELSE
+       /\ scheduler' = res.val
+       /\ IF allBlocksProcessed(scheduler') THEN
+            outEvent' = [type |-> "scFinishedEv", reason |-> "errAfterPeerRemove"]
+          ELSE 
+            UNCHANGED outEvent
+  /\ UNCHANGED scRunning
+           
+handleStatusResponse ==
+  /\ inEvent.type = "bcStatusResponse"
+  /\ LET res == setPeerHeight(scheduler, inEvent.peerID, inEvent.height) IN
+     IF res.err /= noErr THEN
+       /\ outEvent' = [type |-> "scPeerError", peerID |-> inEvent.peerID, reason |-> res.err]
+       /\ UNCHANGED scheduler
+     ELSE
+       /\ scheduler' = res.val
+       /\ outEvent' = noEvent
+  /\ UNCHANGED scRunning
+
+handleTrySchedule == \* every 10 ms, but our spec is asynchronous
+  /\ inEvent.type = "rTrySchedule"
+  /\ LET minH == nextHeightToSchedule(scheduler) IN
+     IF minH = None THEN
+       /\ outEvent' = noEvent
+       /\ UNCHANGED scheduler
+     ELSE IF minH = ultimateHeight+1 THEN
+       /\ outEvent' = noEvent
+       /\ UNCHANGED scheduler
+     ELSE
+       /\ LET hp == highPeers(scheduler, minH) IN
+          IF hp = {} THEN
+            /\ outEvent' = noEvent
+            /\ UNCHANGED scheduler
+          ELSE \E bestPeerID \in hp:
+            /\ LET res == markPending(scheduler, bestPeerID, minH) IN
+               IF res.err /= noErr THEN
+                 outEvent' = [type |-> "scSchedulerFail", error|-> res.err]
+               ELSE
+                 outEvent' = [type |-> "scBlockRequest", peerID |-> bestPeerID, height |-> minH]
+               /\ scheduler' = res.val
+  /\ UNCHANGED scRunning
+
+handleBlockResponse ==
+  /\ inEvent.type = "bcBlockResponse"
+  /\ LET res == markReceived(scheduler, inEvent.peerID, inEvent.height) IN
+     IF res.err /= noErr THEN
```

Comment:
You can maybe extract if then logic to assign newres and newsc and then have single "asignment" (with prime variables).

Reply:
yes, but in general i wanted to make it look as close as possible to the code.